File: src/main/java/com/anfema/ionclient/interceptors/AuthorizationHeaderInterceptor.java
package com.anfema.ionclient.interceptors;


import com.anfema.utils.Log;

import java.io.IOException;
import java.util.concurrent.Callable;

import okhttp3.Interceptor;
import okhttp3.Request;
import okhttp3.Response;

/**
 * Add header 'Authorization' to request.
 */
public class AuthorizationHeaderInterceptor implements Interceptor
{
	private final String           authHeaderValue;
	private final Callable<String> authHeaderValueRetriever;

	/**
	 * Provide value for 'Authorization' header directly.
	 */
	public AuthorizationHeaderInterceptor( String authHeaderValue )
	{
		this.authHeaderValue = authHeaderValue;
		this.authHeaderValueRetriever = null;
	}

	/**
	 * Provide value for 'Authorization' header through the return value of a function.
	 */
	public AuthorizationHeaderInterceptor( Callable<String> authHeaderValueRetriever )
	{
		this.authHeaderValue = null;
		this.authHeaderValueRetriever = authHeaderValueRetriever;
	}

	@Override
	public Response intercept( Chain chain ) throws IOException
	{
		String authHeaderValue = this.authHeaderValue;
		if ( authHeaderValue == null && authHeaderValueRetriever != null )
		{
			try
			{
				authHeaderValue = authHeaderValueRetriever.call();
			}
			catch ( Exception e )
			{
				Log.ex( e );
			}
		}
		return requestWithAuthHeader( authHeaderValue, chain );
	}

	private static Response requestWithAuthHeader( String authHeaderValue, Chain chain ) throws IOException
	{
		Request newRequest = chain.request().newBuilder()
				.addHeader( "Authorization", authHeaderValue )
				.build();
		return chain.proceed( newRequest );
	}
}

File: src/main/java/com/anfema/ionclient/interceptors/AuthorizationHeaderInterceptor.kt
package com.anfema.ionclient.interceptors

import com.anfema.ionclient.utils.IonLog
import okhttp3.Interceptor
import okhttp3.Response
import java.io.IOException

/**
 * Adds header 'Authorization' to request if a value is provided.
 */
class AuthorizationHeaderInterceptor(
    private val authHeaderValueProvider: () -> String?,
) : Interceptor {

    @Throws(IOException::class)
    override fun intercept(chain: Interceptor.Chain): Response {
        val authHeaderValue = authHeaderValueProvider()
        val newRequest = chain.request().newBuilder().apply {
            if (authHeaderValue != null) {
                addHeader("Authorization", authHeaderValue)
            } else {
                IonLog.e(
                    javaClass.simpleName,
                    "No Authorization header was added to request ${chain.request().url}"
                )
            }
        }.build()
        return chain.proceed(newRequest)
    }
}
